rainier-core/src/main/scala/com/stripe/rainier/core/Likelihood.scala

```diff
@@ -4,42 +4,68 @@ import com.stripe.rainier.compute._
 import scala.collection.mutable.ArrayBuffer
 
 trait Likelihood[T] {
-  private[core] type P
-  private[core] def wrapping: Wrapping[T, P]
-  private[core] def logDensity(value: P): Real
+  def real: Real
+  def placeholders: List[Variable]
+  def extract(t: T): List[Double]
 
-  def target(value: T): Target =
-    Target(logDensity(wrapping.wrap(value)))
-
-  def sequence(seq: Seq[T]): Target = {
-    val ph = wrapping.placeholder(seq)
-    val real = logDensity(ph.value)
-    val variables = ph.variables(Nil)
+  def fit(seq: Seq[T]): RandomVariable[Unit] = {
     val arrayBufs =
-      variables.map { _ =>
+      placeholders.map { _ =>
         new ArrayBuffer[Double]
       }
     seq.foreach { t =>
-      val doubles = ph.extract(t, Nil)
+      val doubles = extract(t)
       arrayBufs.zip(doubles).foreach {
         case (a, d) => a += d
       }
     }
     val placeholdersMap =
-      variables.zip(arrayBufs.map(_.toArray)).toMap
-    new Target(real, placeholdersMap)
+      placeholders.zip(arrayBufs.map(_.toArray)).toMap
+    val target = new Target(real, placeholdersMap)
+    new RandomVariable((), Set(target))
   }
 }
 
-object Likelihood {
-  /*
-    We define `fit` in the indirect way below so that the return value of, for example,
-    `Normal(1).fit(...)` is `RandomVariable[Normal]` instead of
-    `RandomVariable[Likelihood[Double]]`. This solves a similar problem to f-bounded
-    polymorphism but without introducing an ugly extra type param.
-   */
-  implicit class Ops[T, L](lh: L)(implicit ev: L <:< Likelihood[T]) {
-    def fit(value: T): RandomVariable[L] = RandomVariable.fit(lh, value)
-    def fit(seq: Seq[T]): RandomVariable[L] = RandomVariable.fit(lh, seq)
+trait LikelihoodMaker {
+  type M[T, P]
+
+  def maker[T, P](implicit ph: Encoder[T, P]): M[T, P]
+  def fromInt = maker[Int, Real]
+  def fromDouble = maker[Double, Real]
+  def fromIntPair = maker[(Int, Int), (Real, Real)]
+  def fromDoublePair = maker[(Double, Double), (Real, Real)]
+  def fromIntVector(size: Int) =
+    maker[Seq[Int], Seq[Real]](Encoder.vector(size))
+  def fromDoubleVector(size: Int) =
+    maker[Seq[Double], Seq[Real]](Encoder.vector(size))
+}
+
+object Likelihood extends LikelihoodMaker {
+  class Maker[T, P](ph: Encoder[T, P]) {
```

Comment:
I don't totally see what these Maker and Fitter classes are doing and it's a little confusing to have slightly different classes with the same names in the Likelihood and Predictor objects. Could these be functions? Or implicit classes where `Maker[T, P]` wraps an `Encoder[T, P]` and has `toLikelihood` and `toPredictor` methods and `Fitter[T, P]` wraps an `Encoder[T,P]` that return a `RandomVariable[Unit]` or a `RandomVariable[Predictor[_,_]]` given the right things?

Reply:
Hmmm. I'll see what I can think of here to simplify and make more clear. The scala type inferencer is a very picky customer.